Smooth Movement 5: God's Dumbest Code Edition (GOLD)

LemonInTheDark opened the pull request:
## About The Pull Request

Hey! I'm back on my bullshit.
This pr was born out of the nausea reported during testing of #85491. In the process of working out why gliding stuttered so much I discovered 2 major issues with our movement visuals. I'll describe them in order, alongside their solutions, and then get into the fun bonuses here.

### Hey wait a minute this sucks

Ok so like, base principles, the idea of gliding is you tell the client "hey this mob is moving, animate it moving from A to B at X pixels per tick".  This is fine, except we don't want mobs to stop and start when moving in game, and if the game lags we will end up delaying their next move by X ms.

Easy solution then, track the average lag and multiply glide rates to correct them.
This is fine, and about what I'm doing, it's just our old method of getting "average lag" was to, once every 10 seconds, get the ratio between real and game time over the PRECEDING 10 seconds. 

Two problems here:
1. We're working off REALLY outdated information, so we're likely to finish early and just sit on the tile for a few game ticks, or worse jump forward because we're moving too slow. This feels jerky and happens CONSTANTLY.
2. Because this is essentially a raw average, we don't care if the lag is sustained (the thing that matters) or due to 1/2 lag spikes (say, hard deletes). This, like the above, would cause us to underestimate how fast it takes for the average next game tick to come out, leading to jumping. This also happens all the time and feels horrid.

Ok so we need better numbers.
Byond provides a measure of the % of the tick the last gametick needed to finish, which would be perfect for this. It's unfortunately averaged in a constant size 16 moving average, so we have to unroll it first. This works, though it does introduce compounding floating point error, since we need to work off the starting values. I've used map_cpu and tick_usage to clamp this error but it's still gonna be there a little bit. Non ideal but fine for this usecase.

We can then track cpu usage over the past X ticks, and then remove outliers and build an average we can use to multiply against new glides. I am decently happy with this though I could use some pointers on the averaging math.

Of note: I tried glide correction, but the jerking and slowdown felt worse/the same as the impacts of lag spikes normally, and it led to extra complexity/cost. I am not 100% confident in my implementation so perhaps there's stuff to do here.

This all works pretty damn good, but it doesn't entirely solve stuttering, even in cases without overtime/client fps drops.

### Oh my god BYOND is stupid

I noticed (alongside @SmArtKar) that the game is FAR smoother before roundstart then after, even if we aren't going into overtime. I was curious about why, and after MUCH digging around, realized that subsystems consuming cpu caused the stuttering.

This was fucking weird. looked more into it, the stutters didn't happen at the same time for all mobs, though they did seem to be roughly consistent across clients maybe? Did a bunch of frame by frame work, died due to my burning hate for this community, came back to life, did even more frame by frame work, and discovered that what was going on was VARIATION in cpu usage dropping frames.

This isn't an us issue, it's an ENGINE issue https://www.byond.com/forum/post/2959355. It occurs locally and remotely, on ss13 code and on raw byond. 

What's going on is (to my understanding didn't fully get lummy's explanation) if the time BETWEEN SendMaps() calls is too different, we may end up skipping client frames ([Background Reading](https://github.com/tgstation/tgstation/blob/master/.github/guides/TICK_ORDER.md)). See the image below.

![art](https://github.com/user-attachments/assets/ff087463-0d3d-4a67-bbea-de04313fe471)

Lummy seemed to be unsure of if he was going to be able to fix this (something about glide compensating) so the only option left to us outside of Deal With ItTM is to ensure the gap between visual data getting sent is as consistent as possible.

This means we need to consume up to the highest cpu values allowed by the MC each tick, since any lower and a spike in work could cause (something close to) a frame skip.

So I am uh, doin that. I have also gone ahead and reserved 4% of the tick for verbs, since we don't really have much control of their cost, and if we reserve 2% for byond I want to have a decent buffer to ensure we don't just force overtime constantly.

This does mean we are limiting the amount of cpu available post sleeping procs. A spike in maptick or verbs cost outside expected bounds then risks pushing us into overtime. I'm not sure of how big a deal this is, thoughts would be useful. The whole idea is really stupid but that's life.

Here's a demonstration of the behavior, to prove my bullshit

https://github.com/user-attachments/assets/dc31b643-96c2-4163-8a0b-5bd12e3f9015

### Debug Tooling

You'll notice a bunch of shit on the top left in the above video, as a part of working this shit out I also added some debug tools.
There's a new verb that enables cpu debug maptext, which displays the tick number and total cpu, sleeping proc, maptick, and verb cost (both live, minimum and maximum over a window). 

I've also added some global values that allow for modifying cpu floor/ceiling and spiking for testing lag.

### Why Is This Drafted?

I need to make sure the error clamping code does its job correctly, otherwise my cpu deaveraging scheme is a non starter. I also need to test to make sure the verb allocation is sufficient for highpop, I don't want to force the game into constant overtime.

## Why It's Good For The Game

The game should be smooth. Duh. Also I can't play/code anymore because the jerkyness makes me depressed.

## Changelog
:cl:
fix: The game should be significantly smoother, both movement gliding and general consistent frame times
/:cl:

LemonInTheDark committed:
Implements better (maybe) lag compensation for gliding

Our existing lag compensation is utter dogwater, it works off a 10
second average that was taken between 0 and 10 seconds ago. Overshooting
leads to stutter steps, which are highly unpleasent, and easy to achive
since any spike in overtime will lead to an overshoot.

What I'm doing here is injesting cpu_usage, deaveraging it using a
world/Tick override (it has an 16 frame moving avg), then taking an
average of that and dropping noisy bits.

I am not the best at stats so this implementation might kinda suck, but
it's better then what we have now.

I'm also trying out spike correction, IE accounting for the delay a
lagspike causes.

This is less of an issue then jumping is because you just end up staying
on a turf for a few ticks longer, but if we cna just slow people down
some I think it's worthwhile.

I am less confident in this concept/implementation then I am with the
tick correction, but we'll see.

I'm putting this up to test on live/motion sick people, should not be
merged.

LemonInTheDark committed:
what the fuck

LemonInTheDark committed:
size increase

LemonInTheDark committed:
Changes avg method to keep a fix sized window

I hate byond

LemonInTheDark committed:
whoop

LemonInTheDark committed:
IT WORKED!!!

LemonInTheDark committed:
oh whoops\

LemonInTheDark committed:
Lemon's amazing magic mob mover

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into dynamic-gliding

LemonInTheDark committed:
Implements potentially better correction and debugging

Glide corrects as if we are a tick in the past (since our info is a tick
out of date), draws more debug info (and draws it to a hidden plane
master) and locks active correction behind a global value

LemonInTheDark committed:
Better debug + glide text

LemonInTheDark committed:
Fixes people seeing my shit, better control, more complete tick readout to shittest

LemonInTheDark committed:
fuck

LemonInTheDark committed:
Noise control, verb

Floors real cpu values with the tick usage/map cpu from the last tick
This serves as a worst case minimum and should bound any error
introduced by floating point bullshit to a particular bound (since error
compounds here)

Adds a verb that toggles on/off the cpu display, so players are not
subjected to it for no reason

LemonInTheDark committed:
Reserves more time for verbs, adds more tracking, autocorrection mode

Verbs take an undefined amount of time (we should really track/manage
this better) and if we're gonna be consuming 100% of the tick we should
limit that (more on that later)

We have map cpu, tick cost, and deaveraged cpu, so we can work out
ROUGHLY how much time we spend on verbs + other bullshit. Should display
that

Adds a new global that enables cpu stabailization, queuing up
/world/Tick() to consume a constant amount of cpu (since that stabalizes
the deltas between sendmaps calls, reducing signficantly visible lag)

LemonInTheDark committed:
Rips out all the testing stuff, proper plane master support

LemonInTheDark committed:
Reorgs somewhat

LemonInTheDark committed:
whoop

LemonInTheDark committed:
whoop2

LemonInTheDark committed:
missed a spot

LemonInTheDark committed:
fuck

Kylerace commented on code/game/world.dm:
```diff
@@ -139,6 +139,216 @@ GLOBAL_VAR(tracy_log)
+	var/non_zero = 0
+	#warn there's gotta be a better way then this, or at least a less shitty way
+	for(var/value in cpu_values)
+		sum += max(value, 100)
```
why clip it?

LemonInTheDark replied:
to not drag the average down, since what we actually want to average here is the LAG, rather then the %

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into the-highway-or-the-hangman

LemonInTheDark committed:
pain

LemonInTheDark committed:
yeet

LemonInTheDark committed:
lint?

LemonInTheDark committed:
whoop

LemonInTheDark committed:
whoop

JohnFulpWillard commented on code/modules/client/client_procs.dm:
```diff
@@ -248,6 +248,8 @@ GLOBAL_LIST_INIT(blacklisted_builds, list(
 
 	GLOB.clients += src
 	GLOB.directory[ckey] = src
+	if(ckey == "lemoninthedark")
```
```suggestion
	if(ckey == "lemoninthedark" || ckey == "johnwillard")
```
hai

LemonInTheDark replied:
I mean whatever man

carlarctg replied:
```suggestion
	if(ckey == "lemoninthedark" || ckey == "johnwillard" || ckey == "carlarc"))
```
me too!!!

LemonInTheDark committed:
whoop

LemonInTheDark commented:
> So the only way to make gliding smooth correctly is to make the game slower?

not slower, we aren't entering overtime, just wasting good electricity.

LemonInTheDark committed:
New form of error correction using the font of truth that is world.cpu

Instead of flooring the cpu values (which allowed spikes to occasionally
just sit undisterbed forever) we correct them as we go using an expanded
world.cpu value. this will not be precise but if each value as we
iterate is corrected then the problem child will always JUST be the new
reading, should constrain error effectively

K4rlox commented:
https://github.com/user-attachments/assets/387d48d6-145a-46bc-9bcb-82599604b8ba

LemonInTheDark committed:
adds control over glide type, more debug shit, canery for negatives

LemonInTheDark committed:
bump er up

LemonInTheDark commented:
just ftr I've had negatives show up on round restart (not server restart). My current working theory is world.cpu persists between rounds, which means we need to carry our list between rounds too! (hhhhh)

SmArtKar commented:
You still working on this?

LemonInTheDark commented:
I intend do but don't have the bandwidth rn (or consistent access to servers for that matter). need to work out the I think? persistence issues and then test on live for a while to make sure that actually worked

nemvar commented:
My theory is that byond just allocates memory for the world.cpu variable, without initializing it. As such, after roundstart it is likely that the process has had already used that memory address and its contents could be anything.
If that's true, world.cpu is undefined for the first 16 gameticks and could contain _any_ value. If that value happens to exceptionally large or small, it can lead to overflows and other nastiness.
Therefore its impossible to perfectly deaverage CPU usage, as the first 16 values will always remain unknown.

No biggi though, just wait with deaveraging until world.time > 16. This introduces some inaccuracy which will luckily decrease over runtime.

Absolucy commented:
@nemvar if this helps, I compiled with this in the latest tg master:
```
/world/Tick()
	log << "time: [time] / cpu: [cpu]"
```
here's the output, up to `world.time = 30`
<details>
<summary>click to expand</summary>

```
Logging in...Fri Feb 14 12:35:12 2025
World opened on network port [removed].
Welcome BYOND! (5.0 Public Version 515.1647)
World loaded at 12:35:20!
Running /tg/ revision: 2025-02-14
HEAD: 24c233ff1cb6e27d3b7d45737c94ec65deb3d0b2
Running rust-g version 3.6.0
Loading config file config.txt...
Loading config file game_options.txt...
Loading config file dbconfig.txt...
Loading config file comms.txt...
Loading config file logging.txt...
Loading config file resources.txt...
Loading config file interviews.txt...
Loading config file lua.txt...
Loading config file auxtools.txt...
Loading config file map_vote.txt...
Loading config file dev_overrides.txt...
Reset configured value for MAPROTATION to original defaults
Reset configured value for TOAST_NOTIFICATION_ON_INIT to original defaults
Reset configured value for ALLOW_RANDOM_EVENTS to original defaults
Reset configured value for HUMANS_NEED_SURNAMES to original defaults
Loading config file maps.txt...
Loading config file word_filter.toml...
Database connection established.
connected
Logged in as Absolucy.
time: 0.5 / cpu: 0
time: 1 / cpu: 37.2137
time: 1.5 / cpu: 37.2224
time: 2 / cpu: 37.536
time: 2.5 / cpu: 37.5479
time: 3 / cpu: 37.5688
time: 3.5 / cpu: 37.5929
time: 4 / cpu: 37.613
time: 4.5 / cpu: 37.6314
time: 5 / cpu: 37.6504
time: 5.5 / cpu: 37.671
time: 6 / cpu: 37.6903
time: 6.5 / cpu: 37.7104
time: 7 / cpu: 37.732
time: 7.5 / cpu: 37.7494
time: 8 / cpu: 37.771
time: 8.5 / cpu: 37.7918
time: 9 / cpu: 0.597682
time: 9.5 / cpu: 0.606395
time: 10 / cpu: 0.31262
time: 10.5 / cpu: 0.32292
time: 11 / cpu: 0.322757
Initialized Title Screen subsystem within 0 seconds!
Initialized Server Tasks subsystem within 0 seconds!
Initialized Input subsystem within 0 seconds!
Initialized Profiler subsystem within 0 seconds!
Initialized Database subsystem within 0 seconds!
Initialized Blackbox subsystem within 0 seconds!
Initialized Admin Verbs subsystem within 0 seconds!
Initialized Sounds subsystem within 0.24 seconds!
time: 11.5 / cpu: 0.318132
Initialized Instruments subsystem within 0.01 seconds!
time: 12 / cpu: 32.5733
time: 12.5 / cpu: 45.4552
time: 13 / cpu: 58.4836
time: 13.5 / cpu: 77.1741
Initialized Greyscale subsystem within 0.53 seconds!
Initialized Vis contents overlays subsystem within 0 seconds!
Initialized Security Level subsystem within 0 seconds!
time: 14 / cpu: 89.4405
time: 14.5 / cpu: 99.8247
time: 15 / cpu: 99.8266
time: 15.5 / cpu: 99.866
time: 16 / cpu: 99.9203
time: 16.5 / cpu: 99.9674
time: 17 / cpu: 100.088
time: 17.5 / cpu: 100.137
time: 18 / cpu: 100.201
time: 18.5 / cpu: 100.244
time: 19 / cpu: 100.285
time: 19.5 / cpu: 100.325
time: 20 / cpu: 68.1198
time: 20.5 / cpu: 55.2851
time: 21 / cpu: 42.3013
time: 21.5 / cpu: 23.6564
time: 22 / cpu: 11.4387
time: 22.5 / cpu: 1.10044
time: 23 / cpu: 1.14037
time: 23.5 / cpu: 1.1469
time: 24 / cpu: 1.13631
time: 24.5 / cpu: 1.12244
Initialized Achievements subsystem within 1.05 seconds!
Initialized Station subsystem within 0 seconds!
Initialized Quirks subsystem within 0 seconds!
Initialized Reagents subsystem within 0.01 seconds!
Initialized Events subsystem within 0 seconds!
Initialized IDs and Access subsystem within 0 seconds!
Initialized Jobs subsystem within 0 seconds!
Initialized AI movement subsystem within 0 seconds!
Initialized Ticker subsystem within 0 seconds!
Initialized AI Controller Ticker subsystem within 0 seconds!
Initialized AI Behavior Ticker subsystem within 0 seconds!
Initialized Trading Card Game subsystem within 0.01 seconds!
Initialized Unplanned AI Controllers subsystem within 0 seconds!
Initialized Unplanned AI Idle Controllers subsystem within 0 seconds!
Initialized AI Idle Controllers subsystem within 0 seconds!
Initialized AI Idle Behaviors subsystem within 0 seconds!
Loading Runtime Station...
time: 25 / cpu: 1.11616
Loaded Station in 0.1s!
time: 25.5 / cpu: 13.32
time: 26 / cpu: 20.8501
time: 26.5 / cpu: 20.8056
time: 27 / cpu: 33.0128
time: 27.5 / cpu: 45.2555
time: 28 / cpu: 57.461
time: 28.5 / cpu: 69.6935
time: 29 / cpu: 81.908
time: 29.5 / cpu: 94.1554
Initialized Mapping subsystem within 0.94 seconds!
time: 30 / cpu: 106.355
```

</details>

LemonInTheDark commented:
> My theory is that byond allocates memory for the world.cpu variable without initializing it. As such, after roundstart it is likely that the process has had already used that memory address and its contents could be anything. If that's true, world.cpu is undefined for the first 16 gameticks and could contain _any_ value. If that value happens to exceptionally large or small, it can lead to overflows and other nastiness. Therefore its impossible to perfectly deaverage CPU usage, as the first 16 values will always remain unknown.

There are decently common cases where it Just WorksTM, I guess something else could be cleaning up that memory but I doubt it? world.cpu should be the PRODUCT of an internal 16 index array, tho I guess THAT could be fucked up somehow? would expect to see negative starting values bout half the time if so and that isn't the case.

> No biggi though, just wait with deaveraging until world.time > 16. This introduces some inaccuracy which will luckily decrease over runtime.

I'm not familiar enough with error propagation to know. Might be worth a go if the carryover shit doesn't work, would prefer a precise number but I suppose if I want that I should just ask lummy to let me turn off the averaging.
